lib/cadet/assessments/assessments.ex

```diff
@@ -29,20 +29,45 @@ defmodule Cadet.Assessments do
     end
   end
 
-  def user_current_story(%User{id: user_id, role: role}) do
-    if role in @submit_answer_roles do
-      Assessment
-      |> where(is_published: true)
-      |> where([a], not is_nil(a.story))
-      |> where([a], a.open_at <= from_now(0, "second") and a.close_at >= from_now(0, "second"))
-      |> join(:left, [a], s in Submission, s.assessment_id == a.id and s.student_id == ^user_id)
-      |> where([_, s], is_nil(s.id) or s.status == "attempting")
-      |> order_by([a], a.open_at)
-      |> order_by([a], a.type)
-      |> select([a], a.story)
-      |> first()
-      |> Repo.one()
+  def user_current_story(user = %User{}) do
+    {:ok, %{result: story}} =
+      Multi.new()
+      |> Multi.run(:default, fn _ -> {:ok, get_user_story_by_direction(user)} end)
+      |> Multi.run(:result, fn %{default: default_res} ->
+        if default_res,
+          do: {:ok, %{all_attempted: false, story: default_res}},
+          else: {:ok, %{all_attempted: true, story: get_user_story_by_direction(user, :backward)}}
+      end)
+      |> Repo.transaction()
+
+    story
+  end
+
+  @spec get_user_story_by_direction(%User{}, :forward | :backward) :: String.t() | nil
+  def get_user_story_by_direction(%User{id: user_id}, direction \\ :forward)
+      when is_atom(direction) do
+    filter_and_sort = fn query ->
+      case direction do
+        :forward ->
+          query
+          |> where([_, s], is_nil(s.id) or s.status == "attempting")
```

Comment:
Prefer atom `:attempting`

Reply:
Also, if I read #179 correctly, `all_attempted === false` if and only if there is any open assessment whose submission's status is `unattempted` (which in our case is basically the `is_nil(s.id)`). I don't think `:attempting` should be here?

Reply:
1. Atoms don't work here. I think the `where` macro just expands it into some raw SQL query (so ecto_enum library is not involved). All our enums as of now should be queried by their string value.

2. In the call, he said that the `all_attempted` flag is essentially to tell whether the first or second kind of logic is used. I personally feel like `attempting` should be treated the same as `not_attempted` in the sense that the assessment has not been "completed".

@ningyuansg Can you confirm whether `allAttempted` flag should be true or false when status is `attempting`?

Reply:
Clarified with him. It should be `false`.

Reply:
Alright, cool.